x-pack/plugins/lens/public/indexpattern_datasource/to_expression.ts

```diff
@@ -265,15 +269,39 @@ function getExpressionForLayer(
       };
     }
 
-    const allDateHistogramFields = Object.values(columns)
-      .map((column) =>
+    const allDateHistogramFields = columnEntries
+      .map(([, column]) =>
         isColumnOfType<DateHistogramIndexPatternColumn>('date_histogram', column) &&
         !column.params.ignoreTimeRange
           ? column.sourceField
           : null
       )
       .filter((field): field is string => Boolean(field));
 
+    // the combination of Interval with extended bounds + Top values (any combination)
+    // requires partial rows to be enabled
+    const shouldEnablePartialRows =
+      columnEntries.some(([, column]) =>
+        isColumnOfType<TermsIndexPatternColumn>('terms', column)
+      ) &&
+      columnEntries.some(
+        ([, column]) =>
+          isColumnOfType<RangeIndexPatternColumn>('range', column) &&
+          column.params.includeEmptyRows &&
+          column.params.customBounds
+      );
+    // This special function will override partial rows value to empty strings
+    // to make it work for the renderer.
+    const partialRowsOverride: ExpressionAstFunction[] = shouldEnablePartialRows
```

Comment:
As discussed offline, this solution seems the "cleanest" so far: the table values gets overridden only in some specific cases and the rest of the logic remains the same.
It might be arguable that the final result produced makes the empty values for the partial rows fall into the `(empty)` category.